cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1103 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::collections::HashMap;
+use std::fmt::{Display, Formatter};
+
+use cedar_policy_core::ast::{
+    BinaryOp, EntityUID, Expr, ExprKind, Literal, PolicyID, PolicySet, RequestType, UnaryOp, Var,
+};
+use cedar_policy_core::entities::err::EntitiesError;
+use cedar_policy_core::impl_diagnostic_from_source_loc_field;
+use cedar_policy_core::parser::Loc;
+use miette::Diagnostic;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use smol_str::SmolStr;
+use thiserror::Error;
+
+use crate::ValidationError;
+use crate::{
+    typecheck::{PolicyCheck, Typechecker},
+    types::{EntityRecordKind, Type},
+    ValidationMode, ValidatorSchema,
+};
+
+/// Data structure storing what data is needed
+/// based on the the [`RequestType`].
+/// For each request type, the [`EntityManifest`] stores
+/// a [`RootAccessTrie`] of data to retrieve.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct EntityManifest<T = ()>
+where
+    T: Clone,
+{
+    /// A map from request types to [`RootAccessTrie`]s.
+    #[serde_as(as = "Vec<(_, _)>")]
+    #[serde(bound(deserialize = "T: Default"))]
+    pub per_action: HashMap<RequestType, RootAccessTrie<T>>,
+}
+
+/// A map of data fields to [`AccessTrie`]s.
+/// The keys to this map form the edges in the access trie,
+/// pointing to sub-tries.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+pub type Fields<T> = HashMap<SmolStr, Box<AccessTrie<T>>>;
+
+/// The root of an entity slice.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub enum EntityRoot {
+    /// Literal entity ids
+    Literal(EntityUID),
+    /// A Cedar variable
+    Var(Var),
+}
+
+impl Display for EntityRoot {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            EntityRoot::Literal(l) => write!(f, "{l}"),
+            EntityRoot::Var(v) => write!(f, "{v}"),
+        }
+    }
+}
+
+/// A [`RootAccessTrie`] is a trie describing
+/// data paths to retrieve. Each edge in the trie
+/// is either a record or entity dereference.
+///
+/// If an entity or record field does not exist in the backing store,
+/// it is safe to stop loading data at that point.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct RootAccessTrie<T = ()>
```

Comment:
Another place where it would be nice to describe what the `T` parameter is for in the doc string.  Also on `AccessTrie`, maybe other types.  If the description is long, you can just put a note like "`T`: see note on OtherType"